utils/cli/cli.go

```diff
@@ -92,41 +95,68 @@ func (ctx CliContext) WithHeight(height int64) CliContext {
 	return ctx
 }
 
-func (ctx CliContext) WithFormerHeight() (CliContext, error) {
+func (ctx CliContext) FormerHeight() (int64, error) {
 	node, err := ctx.context.GetNode()
 	if err != nil {
-		return CliContext{}, err
+		return 0, err
 	}
 
 	status, err := node.Status()
+	if err != nil {
+		return 0, err
+	}
+
+	return status.SyncInfo.LatestBlockHeight - 1, nil
+}
+
+func (ctx CliContext) WithFormerHeight() (CliContext, error) {
+	height, err := ctx.FormerHeight()
 	if err != nil {
 		return CliContext{}, err
 	}
 
-	ctx.context = ctx.context.WithHeight(status.SyncInfo.LatestBlockHeight - 1)
+	ctx.context = ctx.context.WithHeight(height)
 
 	return ctx, nil
 }
 
-func (ctx CliContext) QueryStore(key []byte, storeName string) ([]byte, int64, error) {
-	// Try to query row on `height-1` to avoid delay related to waiting of committing block with height + 1.
+func (ctx CliContext) WithHeightFromFlag() (CliContext, error) {
 	if viper.GetBool(FlagPreviousHeight) {
-		ctx, err := ctx.WithFormerHeight()
-		if err != nil {
-			return nil, 0, err
-		}
+		return ctx.WithFormerHeight()
+	}
 
-		res, height, err := ctx.context.QueryStore(key, storeName)
-		if res != nil {
-			return res, height, err
-		}
+	return ctx.WithHeight(0), nil
+}
+
+func (ctx CliContext) QueryStore(key []byte, storeName string) ([]byte, int64, error) {
+	ctx, err := ctx.WithHeightFromFlag()
+	if err != nil {
+		return nil, 0, err
 	}
-	// request on the current height
-	ctx.context = ctx.context.WithHeight(0)
 
 	return ctx.context.QueryStore(key, storeName)
 }
 
+func (ctx CliContext) QueryStoreAtHeight(height int64, key []byte, storeName string) ([]byte, int64, error) {
+	return ctx.context.WithHeight(height).QueryStore(key, storeName)
+}
+
+func (ctx CliContext) QueryRange(startKey []byte, endKey []byte, limit int,
```

Comment:
Do we call this method somewhere? It looks like we always call `QueryRangeWithTotalAndHandleIO` only, don't we?